### PR TITLE
Core: Don't freeze application if lock file already exists

### LINKED ISSUE #17576: FreeCAD freezes if lockfile already exists
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Problem description

see https://github.com/flathub/org.freecad.FreeCAD/issues/4 and https://github.com/flathub/org.freecad.FreeCAD/issues/5 for whatever reason when running as flatpak `QCoreApplication::applicationPid()` always returns 2 so if you start a second instance of freecad it tries to create an existing lockfile and freezes up. There could be two somewhat separate problems here:

* freecad probably shouldn't freeze because a lockfile exists, in any case it should raise an error? not sure
* using PID is clearly not unique enough in flatpak environment and possibly other situations, maybe we should generate an UUID early in startup and store it for later use instead of the PID. I found 3 uses of `QCoreApplication::applicationPid()`, I don't know if we use PID somewhere else through a different function.
https://github.com/FreeCAD/FreeCAD/blob/b6cda4309a9031bebffa73b694e9a9cc125d64de/src/App/Document.cpp#L944
https://github.com/FreeCAD/FreeCAD/blob/b6cda4309a9031bebffa73b694e9a9cc125d64de/src/Gui/Application.cpp#L2177
https://github.com/FreeCAD/FreeCAD/blob/b6cda4309a9031bebffa73b694e9a9cc125d64de/src/Gui/DocumentRecovery.cpp#L657

### Full version info

```shell
OS: KDE Flatpak runtime (KDE/plasmax11)
Word size of FreeCAD: 64-bit
Version: 1.0.0RC2.38806 (Git)
Build type: Release
Branch: (HEAD detached at 3d63fc6)
Hash: 3d63fc6c2f665a8d5e6468845a419bcac80756c7
Python 3.11.10, Qt 6.7.3, Coin 4.0.3, Vtk 9.2.6, OCC 7.8.1
Locale: C/Default (C)
Stylesheet/Theme/QtStyle: unset/FreeCAD Classic/breeze
```

### Subproject(s) affected?

Core

### Anything else?

_No response_

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

### PR DESCRIPTION
**Pull request by wwmayer**
Fixes #17576

### CODE REVIEW

**Comment by adrianinsaval** [comment]
I have tested this on flatpak added as a patch, is it expected that the app quits if lock can't be acquired? Still this is better than freezing with no information for the user
```
FreeCAD 1.0.0, Libs: 1.0.0RC2R38806 (Git)
(C) 2001-2024 FreeCAD contributors
FreeCAD is free and open-source software licensed under the terms of LGPL2+ license.

Qt: Session management error: Could not open network socket
Failed to create a file lock for the IPC.
The application will be terminated
```

File: src/App/Application.cpp [hunk 1]
```diff
@@ -1114,6 +1115,11 @@ Application::TransactionSignaller::~TransactionSignaller() {
     }
 }
 
+int64_t Application::applicationPid()
+{
+    return QCoreApplication::applicationPid();
+}
```
Comment (adrianinsaval) [inline src/App/Application.cpp:1121]: If I instead use this I can launch two flatpak instances, would that be sensible?
```suggestion
int64_t Application::applicationPid()
{
    static int64_t randomNumber = []() {
        auto seed = std::chrono::high_resolution_clock::now().time_since_epoch().count();
        std::mt19937 generator(static_cast<unsigned>(seed));
        std::uniform_int_distribution<int> distribution(1, 1000000);
        return distribution(generator);
    }();
    return randomNumber;
}
```

**Comment by wwmayer** [comment]
> is it expected that the app quits if lock can't be acquired?

Yes, this is the desired behaviour. This is because two independent FreeCAD instances should not share a common lock file because this can lead to loss of data.
Assuming sharing a common lock file would be supported then this can happen:
1. Start a FreeCAD instance the creates a lock file
2. Start another FreeCAD instance that uses the same lock file
3. Close one of the instances. The lock file will be deleted
4. After modelling something complex FreeCAD crashes
5. When restarting FreeCAD the recovery dialog will not appear because there is no existing lock file

As said in the ticket it's not important to use the application PID from Qt. Important is only to have a unique number for every FreeCAD instance. So, creating a random but static number will be sufficient.

Does a flatpak build use a specific pre-processor define?

File: src/App/Application.cpp (hunk 1 above)
Comment (wwmayer) [reply src/App/Application.cpp:1121]: I would do this in a second PR, then. Although it should work it still requires some testing.

**Comment by yorikvanhavre** [comment]
backported

**Pull request merged**